File: src/main/java/me/zeroeightsix/kami/module/modules/misc/BlockData.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.module.Module;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraftforge.fml.common.gameevent.InputEvent;
import org.lwjgl.input.Mouse;

import java.util.Objects;

import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;

/**
 * @author 0x2E | PretendingToCode
 *
 * TODO: Fix delay timer because that shit broken
 */
@Module.Info(name = "BlockData", category = Module.Category.MISC, description = "Right click blocks to display their data")
public class BlockData extends Module {
    private int delay = 0;

    @Override
    public void onUpdate() {
        if (delay > 0) {
            delay--;
        }
    }

    @EventHandler
    public Listener<InputEvent.MouseInputEvent> mouseListener = new Listener<>(event -> {
        if (Mouse.getEventButton() == 1 && delay == 0) {
            if (mc.objectMouseOver.typeOfHit.equals(RayTraceResult.Type.BLOCK)) {
                BlockPos blockpos = mc.objectMouseOver.getBlockPos();
                IBlockState iblockstate = mc.world.getBlockState(blockpos);
                Block block = iblockstate.getBlock();

                if (block.hasTileEntity()) {
                    TileEntity t = mc.world.getTileEntity(blockpos);
                    NBTTagCompound tag = new NBTTagCompound();
                    Objects.requireNonNull(t).writeToNBT(tag);

                    sendChatMessage(getChatName() + "&6Block Tags:\n" + tag + "");
                }
            }
        }
    });
}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/BlockData.kt
package me.zeroeightsix.kami.module.modules.misc

import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.util.TimerUtils
import me.zeroeightsix.kami.util.text.MessageSendHelper
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.RayTraceResult
import net.minecraftforge.fml.common.gameevent.InputEvent
import org.kamiblue.event.listener.listener
import org.lwjgl.input.Mouse

@Module.Info(
        name = "BlockData",
        category = Module.Category.MISC,
        description = "Right click blocks to display their data"
)
object BlockData : Module() {
    private val timer = TimerUtils.TickTimer()
    private var lastPos = BlockPos.ORIGIN

    init {
        listener<InputEvent.MouseInputEvent> {
            if (Mouse.getEventButton() != 1 || mc.objectMouseOver == null || mc.objectMouseOver.typeOfHit != RayTraceResult.Type.BLOCK) return@listener
            if (timer.tick(5000L) || mc.objectMouseOver.blockPos != lastPos && timer.tick(500L)) {
                val blockPos = mc.objectMouseOver.blockPos
                val blockState = mc.world.getBlockState(blockPos)
                lastPos = blockPos

                if (blockState.block.hasTileEntity(blockState)) {
                    val tileEntity = mc.world.getTileEntity(blockPos) ?: return@listener
                    val tag = NBTTagCompound().apply { tileEntity.writeToNBT(this) }
                    MessageSendHelper.sendChatMessage("""$chatName &6Block Tags:$tag""".trimIndent())
                }
            }
        }
    }
}